File: 2025/src/day03/day03.exs
defmodule Day03 do
  @path "src/day03/input.txt"

  def main do
    raw_file = File.read!(@path)
    banks = parse_joltages(raw_file)
    part1(banks) |> IO.inspect(label: "Part 1: finding max joltages for 2 batteries")
    part2(banks) |> IO.inspect(label: "Part 2: finding max joltages for 12 batteries")
  end

  def part1(banks) do
    banks
    |> Enum.map(&find_max_n_joltages(&1, 2))
    |> Enum.map(fn {acc, _r} -> acc end)
    |> Enum.sum()
  end

  def part2(banks) do
    banks
    |> Enum.map(&find_max_n_joltages(&1, 12))
    |> Enum.map(fn {acc, _r} -> acc end)
    |> Enum.sum()
  end

  defp find_max_n_joltages(bank, batteries) do
    Enum.reduce((batteries - 1)..0//-1, {0, bank}, fn place, {acc, rem} ->
      {max_value, max_index} =
        rem
        |> Enum.with_index()
        |> Enum.take(length(rem) - place)
        |> Enum.max_by(fn {j, _i} -> j end)

      place = max_value * Integer.pow(10, place)
      new_rem = rem |> Enum.drop(max_index + 1)

      {acc + place, new_rem}
    end)
  end

  def parse_joltages(string) do
    string
    |> String.split("\n", trim: true)
    |> Enum.map(fn bank -> String.split(bank, "", trim: true) end)
    |> Enum.map(fn bank ->
      Enum.map(bank, &String.to_integer/1)
    end)
  end
end

Day03.main()
ExUnit.start()

defmodule Day03Test do
  use ExUnit.Case, async: true

  @example_input """
  987654321111111
  811111111111119
  234234234234278
  818181911112111
  """

  test "part1" do
    result =
      @example_input |> Day03.parse_joltages() |> Day03.part1()

    assert result == 357
  end

  test "part2" do
    result =
      @example_input |> Day03.parse_joltages() |> Day03.part2()

    assert result == 3_121_910_778_619
  end
end

File: 2025/src/day03/day03.rs
use std::fs;

const INPUT_PATH: &str = "src/day03/input.txt";

fn main() -> (u64, u64) {
    let raw_file = fs::read_to_string(INPUT_PATH).unwrap();
    let banks = parse_joltages(&raw_file);

    let part1 = part1(banks.clone());
    println!("Part 1: finding max joltages for 2 batteries: {}", part1);
    let part2 = part2(banks.clone());
    println!("Part 2: finding max joltages for 12 batteries: {}", part2);

    (part1, part2)
}

fn part1(banks: Vec<Vec<u32>>) -> u64 {
    banks
        .into_iter()
        .map(|bank| find_max_n_joltages(bank, 2).0)
        .sum()
}

fn part2(banks: Vec<Vec<u32>>) -> u64 {
    banks
        .into_iter()
        .map(|bank| find_max_n_joltages(bank, 12).0)
        .sum()
}

fn parse_joltages(raw_file: &str) -> Vec<Vec<u32>> {
    raw_file
        .lines()
        .map(|bank| bank.chars().map(|b| b.to_digit(10).unwrap()).collect())
        .collect()
}

fn find_max_n_joltages(bank: Vec<u32>, batteries: usize) -> (u64, Vec<u32>) {
    (0..batteries)
        .rev()
        .fold((0u64, bank.to_vec()), |(acc, rem), place| {
            let searchable_len = rem.len() - place;
            let (max_index, &max_value) = rem[..searchable_len]
                .iter()
                .enumerate()
                .reduce(
                    |(max_i, max_v), (i, v)| {
                        if v > max_v { (i, v) } else { (max_i, max_v) }
                    },
                )
                .unwrap();

            let new_place = (max_value as u64) * 10u64.pow(place as u32);
            let new_rem = rem[(max_index + 1)..].to_vec();

            (acc + new_place, new_rem)
        })
}

#[cfg(test)]
mod tests {
    use super::*;

    const EXAMPLE_INPUT: &str = "987654321111111
811111111111119
234234234234278
818181911112111";

    #[test]
    fn test_day03_main() {
        let (part1, part2) = main();
        assert_eq!(part1, 17324);
        assert_eq!(part2, 171846613143331);
    }

    #[test]
    fn test_day03_part1() {
        let banks = parse_joltages(EXAMPLE_INPUT);
        let result = part1(banks);
        assert_eq!(result, 357);
    }

    #[test]
    fn test_day03_part2() {
        let banks = parse_joltages(EXAMPLE_INPUT);
        let result = part2(banks);
        assert_eq!(result, 3_121_910_778_619);
    }
}
